crates/bevy_ecs/src/observer/runner.rs

```diff
@@ -0,0 +1,226 @@
+            descriptor: Default::default(),
+        }
+    }
+}
+
+impl ObserverState {
+    /// Adds the given `event`
+    pub fn with_event(mut self, event: ComponentId) -> Self {
+        self.descriptor.events.push(event);
+        self
+    }
+
+    /// Adds the given `events`
+    pub fn with_events(mut self, events: impl IntoIterator<Item = ComponentId>) -> Self {
+        self.descriptor.events.extend(events);
+        self
+    }
+
+    /// Watches the given [`Entity`] list.
+    pub fn with_entities(mut self, entities: impl IntoIterator<Item = Entity>) -> Self {
+        self.descriptor.entities.extend(entities);
+        self
+    }
+
+    /// Watches the given [`ComponentId`] list.
+    pub fn with_components(mut self, components: impl IntoIterator<Item = ComponentId>) -> Self {
+        self.descriptor.components.extend(components);
+        self
+    }
+}
+
+impl Component for ObserverState {
+    const STORAGE_TYPE: StorageType = StorageType::SparseSet;
+
+    fn register_component_hooks(hooks: &mut ComponentHooks) {
+        hooks.on_add(|mut world, entity, _| {
+            world.commands().add(move |world: &mut World| {
+                world.register_observer(entity);
+            });
+        });
+        hooks.on_remove(|mut world, entity, _| {
+            let descriptor = std::mem::take(
+                &mut world
+                    .entity_mut(entity)
+                    .get_mut::<ObserverState>()
+                    .unwrap()
+                    .as_mut()
+                    .descriptor,
+            );
+            world.commands().add(move |world: &mut World| {
+                world.unregister_observer(entity, descriptor);
+            });
+        });
+    }
+}
+
+/// Type for function that is run when an observer is triggered.
+/// Typically refers to the default runner that runs the system stored in the associated [`ObserverSystemComponent`],
+/// but can be overridden for custom behaviour.
+pub type ObserverRunner = fn(DeferredWorld, ObserverTrigger, PtrMut);
+
+/// An [`Observer`] system. Add this [`Component`] to an [`Entity`] to give it observer behaviors for the given system.
+pub struct Observer<T: 'static, B: Bundle> {
+    system: BoxedObserverSystem<T, B>,
+    descriptor: ObserverDescriptor,
+}
+
+impl<E: Event, B: Bundle> Observer<E, B> {
+    /// Creates a new [`Observer`], which defaults to a "global" observer.
+    pub fn new<M>(system: impl IntoObserverSystem<E, B, M>) -> Self {
+        Self {
+            system: Box::new(IntoObserverSystem::into_system(system)),
+            descriptor: Default::default(),
+        }
+    }
+
+    /// Observe the given `entity`.
+    pub fn with_entity(mut self, entity: Entity) -> Self {
+        self.descriptor.entities.push(entity);
+        self
+    }
+
+    /// Observe the given `entity`. Note that if this is called _after_ an [`Observer`] is spawned, it will produce no effects.
+    pub fn watch_entity(&mut self, entity: Entity) {
```

Comment:
I can imagine use cases where I'd want to add entities to an already spawned observer. Is there any way to do this?

Reply:
`with_entity` uses the builder pattern, which doesn't play very nicely with loops. Given that one of the main use cases of direct Observer spawning is "reusing across a group of entities", I figured it was prudent to include this.